pmlc/dialect/tile/builder.cc

```diff
@@ -769,6 +769,10 @@ TileBuilder::MakeProgram(StringRef name, const ProgramMutations &mutations,
   for (unsigned i = 0; i < returnOp.getNumOperands(); i++) {
     auto userValue = outputs[i];
     auto finalValue = returnOp.getOperand(i);
+    if (!finalValue.getDefiningOp()) {
+      IVLOG(2, "reached condition where return operand has no defining op");
+      userValue = MakePrimitiveOp("ident", {outputs[i]});
+    }
```

Comment:
This is a great solution if it works. Could we remove the special handling at the top of `MakeProgram` now that we have this?

Reply:
Actually, we might need to say:
```
finalValue = MakePrimitiveOp("ident", {userValue});
```

Reply:
Thanks for the suggestion. I tried this, and it still doesn't seem to be affecting the behavior of the builder:

```
2020-05-21 17:56:14,593 VERBOSE-5 [default] ReshapeOp::fold
2020-05-21 17:56:14,594 VERBOSE-2 [default] reached condition where return operand has no defining op
2020-05-21 17:56:14,594 VERBOSE-2 [default] TileBuilder::MakeProgram>


module {
  func @reshape_fold(%arg0: tensor<3x3xsi32> {tile.name = "A"}) -> tensor<3x3xsi32> {
    return %arg0 : tensor<3x3xsi32>
  }
}
```

Reply:
Oh, we can't use `MakePrimitiveOp`, we need to create it directly. This is because there are actually 2 modules here. One is the 'workspace' module that the EDSL directly inserts new ops into. But by the end of `MakeProgram` we've actually cloned a subgraph based on the trace into a new module. We need to insert the `ident` into this new module.

Reply:
I tried this and it segfaulted. Perhaps I'm misunderstanding what I need to do here?

```
    if (!finalValue.getDefiningOp()) {
      IVLOG(2, "reached condition where return operand has no defining op");
      auto ident = builder.create<eltwise::IdentOp>(impl->loc, userValue);
      module.push_back(ident);
    }
```

Reply:
No need to do `module.push_back(ident)` (this would effectively add an op at the top-level, outside of the `FuncOp`). When you use `builder.create`, this should create the op in the location of the current insertion point. So you'll likely need to adjust the insertion point. I would actually create a new `OpBuilder` and use the constructor that takes an `Operation*` since that ctor will initialize the insertion point to be just before the specified operation. In this case, you'd want to insert this just before `return` op.